programs/validator-history/src/state.rs

```diff
@@ -454,4 +549,53 @@ mod tests {
     fn test_validator_history_layout() {
         println!("{}", ValidatorHistoryEntry::type_layout());
     }
+
+    #[test]
+    fn test_epoch_range() {
+        // Add in 4 CircBuf entries, with epoch 0, 1, 2, 3
+        let mut circ_buf = CircBuf::default();
+        for i in 0..4 {
+            let entry = ValidatorHistoryEntry {
+                epoch: i,
+                ..ValidatorHistoryEntry::default()
+            };
+            circ_buf.push(entry);
+        }
+        // Test epoch range [0, 3]
+        let epoch_range = circ_buf.epoch_range(0, 3).unwrap();
+        assert_eq!(epoch_range.len(), 4);
+
+        // skip an epoch in the middle
+        circ_buf = CircBuf::default();
+        for i in 0..4 {
+            if i == 2 {
+                continue;
+            }
+            let entry = ValidatorHistoryEntry {
+                epoch: i,
+                ..ValidatorHistoryEntry::default()
+            };
+            circ_buf.push(entry);
+        }
+
+        // Test epoch range [0, 3]
+        let epoch_range = circ_buf.epoch_range(0, 3).unwrap();
+        assert_eq!(epoch_range.len(), 3);
+
+        // same start and end epoch
+        let epoch_range = circ_buf.epoch_range(0, 0).unwrap();
+        assert_eq!(epoch_range.len(), 1);
+
+        // Test start epoch out of range
+        let epoch_range = circ_buf.epoch_range(4, 3);
+        assert!(epoch_range.is_none());
+
+        // Test end epoch out of range
+        let epoch_range = circ_buf.epoch_range(0, 5);
+        assert!(epoch_range.is_none());
+
+        // None if start epoch is none
+        let epoch_range = circ_buf.epoch_range(2, 3);
```

Comment:
same here, seems like it should just return 3?

Reply:
there's a chance weird case might be hit. let's say we're doing 20 epoch lookback and miss an epoch. everything will be fine until you hit the missing epoch, at which point epoch_range will return nothing, but there's a bunch of data in the buffer

Reply:
fixed